DataFormats/CTPPSDetId/src/CTPPSDiamondDetId.cc

```diff
@@ -0,0 +1,99 @@
+/****************************************************************************
+ * Author: Seyed Mohsen Etesami 
+ * Spetember 2016
+ ****************************************************************************/
+
+
+#include "DataFormats/CTPPSDetId/interface/CTPPSDiamondDetId.h"
+
+#include "FWCore/Utilities/interface/Exception.h"
+
+using namespace std;
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSDiamondDetId::CTPPSDiamondDetId(uint32_t id) : CTPPSDetId(id)
+{
+  if (! Check(id))
+    {
+      throw cms::Exception("InvalidDetId") << "CTPPSDiamondDetId ctor:"
+					   << " det: " << det()
+					   << " subdet: " << subdetId()
+					   << " is not a valid CTPPS Timing Diamond id";  
+    }
+}
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSDiamondDetId::CTPPSDiamondDetId(uint32_t Arm, uint32_t Station, uint32_t RomanPot, uint32_t Plane, uint32_t Det) :       
+  CTPPSDetId(sdTimingDiamond, Arm, Station, RomanPot)
+{
+  if (Arm > maxArm || Station > maxStation || RomanPot > maxRP || Plane > maxPlane || Det > maxDet)
+    {
+      throw cms::Exception("InvalidDetId") << "CTPPSDiamondDetId ctor:" 
+					   << " Invalid parameters:" 
+					   << " arm=" << Arm
+					   << " station=" << Station
+					   << " rp=" << RomanPot
+					   << " plane=" << Plane
+					   << " detector=" << Det
+					   << std::endl;
+    }
+
+  uint32_t ok=0xfe000000;
+  id_ &= ok;
+
+  id_ |= ((Arm & maskArm) << startArmBit);
+  id_ |= ((Station & maskStation) << startStationBit);
+  id_ |= ((RomanPot & maskRP) << startRPBit);
+  id_ |= ((Plane & maskPlane) << startPlaneBit);
+  id_ |= ((Det & maskDet) << startDetBit);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+std::ostream& operator << (std::ostream& os, const CTPPSDiamondDetId& id)
+{
+  os << "arm=" << id.arm()
+     << " station=" << id.station()
+     << " rp=" << id.rp()
+     << " plane=" << id.plane()
+     << " Detector=" << id.det();
+
+  return os;
+}
+
+
+
+
+
+//----------------------------------------------------------------------------------------------------
+
+string CTPPSDiamondDetId::planeName( NameFlag flag)
+{
+  string name;
+  if (flag == nFull) name = rpName(flag) + "_";
+  if (flag == nPath) name = rpName(flag) + "/plane ";
+
+  char buf[10];
+  uint32_t planeID = plane();
+  sprintf(buf, "%02u", planeID);
+
+  return name + buf;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+string CTPPSDiamondDetId::channelName(NameFlag flag)
```

Comment:
I find this design rather terrible from the performance point of view
- multiple string instantiations are needed to compute the name, all returned by value
- the hierarchy of calls "channel -> plane -> rp -> station -> arm -> subdetector" suggests that while looping over channels the same subdetector value s will have to be recomputed nC*nP*nR*nS*nA times
- a name, which is supposed to be unique to the given channel ID (at worst in two flavors nFull and nPath) is accessible only via a full recomputation of it.
- support of multiple formatting versions is a bad direction for data formats: data formats should change the least and not serve needs of algorithms that may need some pretty printing.

Current use case appears to be only in the DQM (not even in the ```operator << ```). If it's staying in DQM,
then these "xxxName" methods should be just in the DQM.

If some more common and interactive use is needed to be available via data formats:
- since range in each layer of hierarchy is rather small, define static const arrays to get their values and make access methods for them (inlined, returning by reference), so that e.g. "CTPPS/TrackingStrip/sector45/station210/nr_bt/plane1/ch5"  can be multiplexed from separate static storage elements "CTPPS/TrackingStrip", "sector45", "station210", "nr_bt", "plane1", "ch5" (it may be worth to even define a static array for the multiplexed names, as long as the size is about o(1000) )
- inline the individual layer name access and the name-multiplexing methods.